all: replace log15 with slog

jwasinger opened the pull request:
This PR replaces Geth's logger package (a fork of [log15](https://github.com/inconshreveable/log15)) with an implementation using slog, a logging library included as part of the Go standard library as of Go1.21.

Main changes are as follows:
* removes any log handlers that were unused in the Geth codebase.
* Json, logfmt, and terminal formatters are now slog handlers.
* Verbosity level constants are changed to match slog constant values.  Internal translation is done to make this opaque to the user and backwards compatible with existing `--verbosity` and `--vmodule` options.
* `--log.backtraceat` and `--log.debug` are removed.

The external-facing API is largely the same as the existing Geth logger.  Logger method signatures remain unchanged.

A small semantic difference is that a `Handler` can only be set once per `Logger` and not changed dynamically.  This just means that a new logger must be instantiated every time the handler of the root logger is changed.

----
For users of the `go-ethereum/log` module. If you were using this module for your own project, you will need to change the initialization. If you previously did 
```golang
log.Root().SetHandler(log.LvlFilterHandler(log.LvlInfo, log.StreamHandler(os.Stderr, log.TerminalFormat(true))))
```
You now instead need to do 
```golang
log.SetDefault(log.NewLogger(log.NewTerminalHandlerWithLevel(os.Stderr, log.LevelInfo, true)))
```
See more about reasoning here: https://github.com/ethereum/go-ethereum/issues/28558#issuecomment-1820606613

MariusVanDerWijden commented:
The diff already looks promising, do you have some benchmarks for the new logging lib?

jwasinger commented:
I did some benchmarks (https://github.com/ethereum/go-ethereum/blob/bf91275fc6baf3de3028458ad772a6de3dbabd20/log/slog/log/slog_test.go), but the code has since become outdated and I need to update them and add new benchmarks.

```
BenchmarkSlogJsonLogger-12          	  829551	      1398 ns/op
BenchmarkGethJsonLogger-12          	 1000000	      1149 ns/op
BenchmarkSlogTerminalLogger-12      	 1302140	       918.3 ns/op
BenchmarkGethTerminalLogger-12      	 1971841	       610.5 ns/op
```

I think those can be improved on to make slog on par with our current logger in terms of performance.

holiman commented:
> I need to update them and add new benchmarks.

Do you have any newer? Seems not possible to run those. Also, would be interesting to see the performance difference for logging on not-enabled levels, e.g. perf on `TRACE` when we've only enabled `INFO`.

holiman commented:
Looking into it a bit, you have this curious construction where we still use the `ethereum/log` package, and under the hood, it holds an `slog` logger: 

```
func (h *swapLogger) write(level slog.Level, msg string, attrs ...any) {
	(*h.innerLogger.Load().(*slog.Logger)).Log(context.Background(), level, msg, attrs...)
}
```

Could you please add some information to the PR description about the design/implementation choices that you have made, and why?

jwasinger commented:
@holiman updated the description with an explanation.

holiman commented on log/logger.go:
```diff
@@ -1,33 +1,61 @@
+	// TODO: re-incorporate this here instead of inside func
+)
 
+// we use custom error levels here (instead of default provided from slog) to keep backwards
+// compatibility with pre-existing geth log lib
```
Just, in general, how much 'cruft' could we get rid of if we didn't care a whole lot about backwards compatibility? What's your general feeling?

jwasinger replied:
This particular detail is going away in the refactor:

There will be no notion of a separate `Level` in the Geth `log` package.  We will convert internally to match the level values used by slog (e.g. in `Vmodule()`).

jwasinger replied:
Other than that, I'm not sure. I thought a bit about using the built-in handlers (json and text handlers).  But the text handler seems basic (only supports logfmt, not terminal format).

One issue is that these are standalone handlers and not "formatters" (which don't exist in `slog`).  So I would lean to keep the existing formatters because 1) we know they work and 2) we already have a distinction between Handler and Formatter which is kind of nice IMO.

jwasinger replied:
also, because `slog` uses `Record.PC` instead of associating a call stack to the record, I remove the reliance on `go-stack/stack`.  The only downside is that this has some interaction with the `LazyHandler` where it is harder to extract the relevant portion of a call-stack if one is returned from a lazy evaluation (we can't use the utilities provided by `go-stack/stack` to trim [here](https://github.com/ethereum/go-ethereum/blob/4e1e37323d87967708b58546f040990db7c2820c/log/handler.go#L285-L288) if we don't have a `CallStack` object associated with the `Record`, available to use).

Since we don't use this functionality anywhere, I just remove this trimming.

jwasinger replied:
Also I assume that we don't want to keep any previously-built-in handler that we aren't using in the Geth codebase.  So I've removed them.  Users can just re-implement on their own.

holiman commented on log/format.go:
```diff
@@ -156,55 +161,74 @@ func TerminalFormat(usecolor bool) Format {
 	buf.WriteByte('\n')
 }
 
+// TODO: investigate use of built-in slog json handler so we can remove the one here
```
Yes, definitely seems like something that `slog` should support out of the box

jwasinger replied:
It does.  Thinking about it more, it actually should be straightforward to use slog's built-in handlers for logfmt/json formats and keep the existing logfmt in the Geth codebase.  These can be sub-handlers.

holiman commented on log/format_test.go:
```diff
@@ -2,120 +2,12 @@ package log
-		int string
-		s   string
-	}{
-		{"111222333444555678999", "111,222,333,444,555,678,999"},
-		{"-111222333444555678999", "-111,222,333,444,555,678,999"},
```
Have these testcases been lifted over somewhere else? IMO we want to retain this behaviour (and some form of tests for this behaviour)

jwasinger replied:
No.  This reminded me that by default, the slog logfmt/json handlers emit output that is a different format than our current ones.  That's a missing piece that still needs to be added to this PR (as well as reintroducing test coverage for it).

jwasinger commented:
Are we okay with removing the alignment of provided attributes when using `logfmt` format?

`t=2023-10-18T23:44:05+0800 lvl=info msg="NAT mapped port"                       proto=TCP extport=30303 intport=30303 interface="UPNP IGDv1-IP1"` -> `t=2023-10-18T23:44:05+0800 lvl=info msg="NAT mapped port" proto=TCP extport=30303 intport=30303 interface="UPNP IGDv1-IP1"`

This is actually a bit of a hack to implement given the rigidity of [`ReplaceAttr`](https://pkg.go.dev/golang.org/x/exp/slog#HandlerOptions) feature of `HandlerOptions`.

namiloh commented:
> the logfmt format is mostly meant to be consumed by scripts anyways.

I agree. Imo we don't require the alignment

jwasinger commented:
another break in backwards compatibility for logfmt:  we have to print the level as a quoted string.  There's no way around this afaict without ditching the built-in text handler.  Seems like not a huge issue.

jwasinger commented:
I still need to add proper formatting for `Uint256` and `big.Int` types.  But that seems like the only other thing outstanding, and it's a matter of direct copying from the current logger code.

jwasinger commented:
Updated benchmarks from my local machine:

This PR ([source](https://github.com/jwasinger/slog-benchmarks/commit/87fb93db4be56e80c9f3e05e080bff3f59c7e9f0)):
```
BenchmarkJson-12               	 1169445	      1024 ns/op
BenchmarkGloggerDisabled-12    	212140161	         5.626 ns/op
BenchmarkGloggerDiscard-12     	 3455035	       346.0 ns/op
BenchmarkGloggerTerminal-12    	  789492	      1499 ns/op
BenchmarkGloggerLogfmt-12      	  777709	      1515 ns/op
BenchmarkGloggerJson-12        	  801158	      1492 ns/op
```

Current Geth Log15-based logger ([source](https://github.com/jwasinger/slog-benchmarks/commit/b91d507b33096752d23974bd0e0d397ca67c1a0e)):
```
BenchmarkJson-12               	  895046	      1329 ns/op
BenchmarkGloggerDisabled-12    	 9091378	       131.3 ns/op
BenchmarkGloggerDiscard-12     	 9066014	       132.8 ns/op
BenchmarkGloggerTerminal-12    	 1000000	      1029 ns/op
BenchmarkGloggerLogfmt-12      	 1000000	      1105 ns/op
BenchmarkGloggerJson-12        	  714872	      1671 ns/op
```

holiman commented:
Ready for review

holiman commented on cmd/evm/blockrunner.go:
```diff
@@ -49,7 +49,6 @@ func blockTestCmd(ctx *cli.Context) error {
 	if len(ctx.Args().First()) == 0 {
 		return errors.New("path-to-test argument required")
 	}
-
```
please undo

holiman commented on eth/protocols/snap/sync.go:
```diff
@@ -2661,7 +2661,7 @@ func (s *Syncer) OnStorage(peer SyncPeer, id uint64, hashes [][]common.Hash, slo
 		size += common.StorageSize(len(node))
 	}
-	logger := peer.Log().New("reqid", id)
+	logger := peer.Log().With("reqid", id)
```
The `log.Logger` interface is 'ours', if you add `New` then these changes can be undone

```golang
// A Logger writes key/value pairs to a Handler
type Logger interface {
	// Handler returns the handler associated with this logger
	Handler() slog.Handler

	// TODO: add WithGroup()?

	// With returns a new Logger that has this logger's attributes plus the given attributes
	With(ctx ...interface{}) Logger

	// New returns a new Logger that has this logger's attributes plus the given attributes. Identical to With
	New(ctx ...interface{}) Logger
```

holiman commented on p2p/server_test.go:
```diff
@@ -75,7 +75,7 @@ func startTestServer(t *testing.T, remoteKey *ecdsa.PublicKey, pf func(*Peer)) *
 		NoDiscovery: true,
 		PrivateKey:  newkey(),
-		Logger:      testlog.Logger(t, log.LvlTrace),
+		Logger:      testlog.Logger(t, log.LevelTrace),
```
Here too, a lot of changes could be avoided if you just 
```golang

const (
	levelMaxVerbosity slog.Level = math.MinInt
	LevelTrace        slog.Level = -8
	LevelDebug                   = slog.LevelDebug
	LevelInfo                    = slog.LevelInfo
	LevelWarn                    = slog.LevelWarn
	LevelError                   = slog.LevelError
	LevelCrit         slog.Level = 12

	LvlTrace = LevelTrace // for backward-compatibility
````

holiman reviewed (commented):
It would be easier to review if we minimize the non-important changes

holiman commented:
It seems that you haven't actually checked in the benchmarks? Why not add them to the PR? e.g. this could go into `logger_test.go`: 

```golang
func BenchmarkGloggerTerminal(b *testing.B) {
	glogHandler := NewGlogHandler(TerminalHandler(io.Discard, false))
	glogHandler.Verbosity(LevelInfo)
	l := NewLogger(glogHandler)

	var bb = make([]byte, 10)
	for i := 0; i < b.N; i++ {
		binary.BigEndian.PutUint64(bb, uint64(i))
		l.Info("This is a message", "foo", i, "baz", bb, "bonk", "boo")
	}
}
```

holiman commented:
```diff
diff --git a/log/logger.go b/log/logger.go
index 5426425c58..b3f3b9fc5c 100644
--- a/log/logger.go
+++ b/log/logger.go
@@ -32,7 +32,7 @@ const (
 
 	// for backward-compatibility
 	LvlTrace = LevelTrace
-	LvlInfo = LevelInfo
+	LvlInfo  = LevelInfo
 	LvlDebug = LevelDebug
 )
 
@@ -95,7 +95,7 @@ func LevelString(l slog.Level) string {
 	case slog.LevelWarn:
 		return "warn"
 	case slog.LevelError:
-		return "error"
+		return "eror"
 	case LevelCrit:
 		return "crit"
 	default:
```
This makes the `go test -tags=integrationtests ./cmd/geth -run TestLogging` tests nearly pass: remainign issues are 
1. Whitespace-issues which are acceptable (or even: better in this PR than on master)
2. int-issues. In this PR, numbers are not pretty-printed in `logfmt`. E.g.

```
 have t=XXXXXXXXXXXXXXXXXXXXXXXX lvl=info msg=int64 1,000,000=1000000
 want t=XXXXXXXXXXXXXXXXXXXXXXXX lvl=info msg=int64 1,000,000=1,000,000
```
This is also fine, IMO, since `logfmt` is a machine-format. But in the name of consistency, if so we should also _not_ pretty-pirint `uint256.Int` and `big.Int`, which this PR does. 
```
t=2023-10-20T12:56:08+0200 lvl=info msg=uint256 11,122,233,344,455,567,899,900=11,122,233,344,455,567,899,900
```

holiman commented on cmd/evm/internal/t8ntool/transaction.go:
```diff
@@ -24,6 +24,8 @@ import (
 	"strings"
 
+	"golang.org/x/exp/slog"
+
```
```suggestion

```

holiman commented on cmd/faucet/faucet.go:
```diff
@@ -39,6 +39,8 @@ import (
 	"time"
 
+	"golang.org/x/exp/slog"
+
```
drop ws?

holiman commented on internal/debug/flags.go:
```diff
@@ -35,6 +35,8 @@ import (
 	"github.com/mattn/go-isatty"
 	"github.com/urfave/cli/v2"
 	"gopkg.in/natefinch/lumberjack.v2"
+
```
why whitespace?

holiman commented on internal/debug/flags.go:
```diff
@@ -75,17 +77,6 @@ var (
 		Usage:    "Write logs to a file",
 		Category: flags.LoggingCategory,
 	}
-	backtraceAtFlag = &cli.StringFlag{
-		Name:     "log.backtrace",
-		Usage:    "Request a stack trace at a specific logging statement (e.g. \"block.go:271\")",
-		Value:    "",
-		Category: flags.LoggingCategory,
-	}
-	debugFlag = &cli.BoolFlag{
-		Name:     "log.debug",
-		Usage:    "Prepends log messages with call-site location (file and line number)",
-		Category: flags.LoggingCategory,
-	}
```
Don't just remove them, they first should be deprecated: if a user uses them, we should print out "Hey log.backtraceat is deprecated, do not use this option or we will come over there and kick your ass". 

And then they get a release or fifteen to remove the flag from their config, before we actually drop it and refuse to start up due to `flag provided but not defined: -log.backtrace`

holiman commented on internal/debug/flags.go:
```diff
@@ -341,8 +331,8 @@ func StartPProf(address string, withMetrics bool) {
 	Handler.StopGoTrace()
-	if closer, ok := logOutputStream.(io.Closer); ok {
-		closer.Close()
+	if logOutputF != nil {
+		logOutputF.Close()
```
As far as I can tell, nothing ever sets the `logOutputIF` ?

holiman replied:
Hm, the same problem can be said for current `master` too, ahem..

holiman replied:
Perhaps you shound make the `logOuputIF` into an `io.Closer` instead

jwasinger replied:
Can we just remove `logOutputF` and not explicitly close the log file on shutdown?

holiman replied:
Hm.. I think it's prudent to close files if we have them open. It's not that big of a hassle

holiman commented on internal/debug/flags.go:
```diff
@@ -241,27 +212,55 @@ func Setup(ctx *cli.Context) error {
+		fileOutput = &lumberjack.Logger{
 			Filename:   logFile,
 			MaxSize:    ctx.Int(logMaxSizeMBsFlag.Name),
 			MaxBackups: ctx.Int(logMaxBackupsFlag.Name),
 			MaxAge:     ctx.Int(logMaxAgeFlag.Name),
 			Compress:   ctx.Bool(logCompressFlag.Name),
 		}
-		ostream = log.StreamHandler(io.MultiWriter(output, lumberWriter), logfmt)
+		output = io.MultiWriter(terminalOutput, fileOutput)
 	} else if logFile != "" {
-		f, err := os.OpenFile(logFile, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
-		if err != nil {
+		var err error
+		if fileOutput, err = os.OpenFile(logFile, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644); err != nil {
```
the lumberjack.Logger is an `io.Closer`, and so is the `fileOutput`. So in both these cases, you need to set the `logOutputIF` to it

holiman commented on signer/core/auditlog.go:
```diff
@@ -19,6 +19,9 @@ package core
 	"encoding/json"
+	"os"
+
+	"golang.org/x/exp/slog"
```
rm whitespace below

holiman commented on signer/storage/aes_gcm_storage_test.go:
```diff
@@ -23,6 +23,8 @@ import (
 	"testing"
 
+	"golang.org/x/exp/slog"
+
```
rm pls

holiman reviewed (commented):
Looking pretty good

holiman commented on log/format.go:
```diff
@@ -98,212 +73,102 @@ type TerminalStringer interface {
 func LogfmtFormat() Format {
-	return FormatFunc(func(r *Record) []byte {
-		common := []interface{}{r.KeyNames.Time, r.Time, r.KeyNames.Lvl, r.Lvl, r.KeyNames.Msg, r.Msg}
+	return FormatFunc(func(r slog.Record) []byte {
```
So, it seems that the `func LogfmtFormat() Format` is not actually used, and can be just dropped, the _actual_ handler for `logfmt` is constructed thus: 
```golang
func LogfmtHandler(wr io.Writer) slog.Handler {
	return &builtinHandler{slog.NewTextHandler(wr, &slog.HandlerOptions{
		ReplaceAttr: builtinReplace,
	})}
}
```
Curious...

holiman replied:
IMO drop this method

holiman commented on log/handler.go:
```diff
@@ -277,13 +277,13 @@ func builtinReplace(_ []string, attr slog.Attr) slog.Attr {
 			attr.Value = slog.AnyValue("<nil>")
 		} else {
-			attr.Value = slog.AnyValue(formatLogfmtBigInt(v))
+			attr.Value = slog.AnyValue(v.String())
```
It seems to me that for some reason, the case of `*big.Int` is not needed? Not quite sure why. Perhaps `Stringer` ? Try dropping this clause, it still works.

jwasinger replied:
We explicitly return `slog.StringValue` now.

holiman commented on log/handler.go:
```diff
@@ -277,13 +277,13 @@ func builtinReplace(_ []string, attr slog.Attr) slog.Attr {
 			attr.Value = slog.AnyValue("<nil>")
 		} else {
-			attr.Value = slog.AnyValue(formatLogfmtUint256(v))
+			attr.Value = slog.AnyValue(v.ToBig().String())
```
`AnyValue` -> `StringValue`. And use `v.Dec`, so `attr.Value = slog.StringValue(v.Dec())` instead

jwasinger replied:
if we use `StringValue`, the value will be quoted in the output.

jwasinger replied:
RE relevant comment below:  I was mistaken.  To avoid values in the output getting quoted we must set the value to `slog.AnyValue(fmt.Stringer)`.  so we would need to wrap `uint256.Int` in a custom type to make the default string format decimal and not hex.

holiman replied:
Hm. This is what I have
```
	switch v := attr.Value.Any().(type) {
	case time.Time:
		attr = slog.Any(attr.Key, v.Format(timeFormat))
	case *uint256.Int:
		if v == nil {
			attr.Value = slog.StringValue("<nil>")
		} else {
			attr.Value = slog.StringValue(v.Dec())
		}
	}
```
And it spits out this: 
```
$ go run -tags=integrationtests ./cmd/geth --log.format=logfmt logtest 
t=2023-11-08T12:41:23+0100 lvl=info msg=big.Int 111,222,333,444,555,678,999=111222333444555678999
t=2023-11-08T12:41:23+0100 lvl=info msg=-big.Int -111,222,333,444,555,678,999=-111222333444555678999
t=2023-11-08T12:41:23+0100 lvl=info msg=big.Int 11,122,233,344,455,567,899,900=11122233344455567899900
t=2023-11-08T12:41:23+0100 lvl=info msg=-big.Int -11,122,233,344,455,567,899,900=-11122233344455567899900
t=2023-11-08T12:41:23+0100 lvl=info msg=uint256 111,222,333,444,555,678,999=111222333444555678999
t=2023-11-08T12:41:23+0100 lvl=info msg=uint256 11,122,233,344,455,567,899,900=11122233344455567899900

```

No problematic quoting.. ?

jwasinger replied:
no.  I'm completely wrong.  it appears only byte-slices get quoted. https://cs.opensource.google/go/x/exp/+/7918f672:slog/text_handler.go;l=114

holiman commented:
A couple of more cases that we should also cover
```diff
diff --git a/cmd/geth/logtestcmd_active.go b/cmd/geth/logtestcmd_active.go
index c66013517a..69a8d86a49 100644
--- a/cmd/geth/logtestcmd_active.go
+++ b/cmd/geth/logtestcmd_active.go
@@ -130,5 +130,9 @@ func logTest(ctx *cli.Context) error {
 		log.Info("Inserted known block", "number", 99, "hash", common.HexToHash("0x12322"), "txs", 10, "gas", 1, "other", "third")
 		log.Warn("Inserted known block", "number", 1_012, "hash", common.HexToHash("0x1234"), "txs", 200, "gas", 99, "other", "fourth")
 	}
+	{
+		log.Info("nil-big.Int", "<nil>", (*big.Int)(nil))
+		log.Info("nil-uint256.Int", "<nil>", (*uint256.Int)(nil))
+	}
 	return nil
 }
```

holiman commented on log/handler.go:
```diff
@@ -1,375 +1,290 @@
-type swapHandler struct {
-	handler atomic.Value
-}
+func (h *builtinHandler) Handle(_ context.Context, r slog.Record) error {
+	var sanitizedAttrs []slog.Attr
+	sanitizedRecord := slog.NewRecord(r.Time, r.Level, r.Message, r.PC)
```
I think the `buildinHandler` logic is a bit too convoluted. There's this logic to first detect e..g `time`, conflicting with the 'native' key `time`. And so we change it into `xxxxxxtime` here in `Handle`, and then later on in `builtinReplace` we check every attribute: to see if it is `xxxxxxtime`, if so, swap it back to `time`. 

We jump through a lot of hoops just to make it possible for a developer to use certain keys: `time`, `level`, `msg`, `source`. 

IMO it's not that big of a deal if we were to emit e.g.`_time` or `time_0` or even `xxxxtime` on such a conflict (so don't bother replacing back)

holiman replied:
I guess what I really mean is: us devs _should not_ use those keys, because it's ambiguous, and won't work well in in `json`, or any format which is mapping-based. So we should just avoid it. But with such fancy / nice workarounds to make it work anyway, we won't even notice if we ever happen to do this. 

Ideally, we'd detect at compilation time that a reserved keyword was used, but, well, can't do that in go :(

jwasinger replied:
we could also just remove the builtinHandler entirely and make logfmt/json handlers faster for us as a result.  If we (or users) happen to use conflicting keys, they can deal with the consequences (keys appearing twice in the outputted record).

holiman replied:
> we could also just remove the builtinHandler entirely and make logfmt/json handlers faster for us as a result. If we (or users) happen to use conflicting keys, they can deal with the consequences (keys appearing twice in the outputted record).

Yes, I think so too. As you can see in my other PR to master, the current json logger already let's the custom fields overwrite the 'native' ones, so that's already broked IMO. I think the simplicity is worth it

holiman commented on internal/debug/flags.go:
```diff
@@ -77,6 +76,17 @@ var (
+		Usage:    "Request a stack trace at a specific logging statement (e.g. \"block.go:271\")",
+		Value:    "",
+		Category: flags.LoggingCategory,
+	}
```
Move htem into `flags_legacy.go`, where other deprecated flags reside. Belonging to `flags.DeprecatedCategory`.

Also add a note `(deprecated)` to the Usage-string

jwasinger replied:
wondering where would be a proper place to log a warning?  if I put them in `flags_legacy.go`, I cannot import into `internal/debug/flags.go` without creating an import cycle.

holiman commented:
https://cs.opensource.google/go/x/exp/+/master:slog/value.go;l=170

Slog.AnyValue internally returns slog.StringValue

holiman commented on log/format.go:
```diff
@@ -98,212 +73,102 @@ type TerminalStringer interface {
-	})
-}
-
 func formatShared(value interface{}) (result interface{}) {
```
This `formatShared` method is now somewhat moot. I tried to hit it, with the following testcase

```golang
type customError struct{}

func (c *customError) Error() string { return "" }

type customStruct struct {
	A string
}

func TestLogging(t *testing.T) {
	b := new(bytes.Buffer)
	logger := NewLogger(NewGlogHandler(TerminalHandler(b, false)))

	logger.Info("test",
		"nil-stringer-interface", (fmt.Stringer)(nil),
		"nil-error-interface", error(nil),
		"nil-custom-err", (*customError)(nil),
		"nil-concrete-stringer", (*time.Time)(nil),
		"nil-custom-struct", (*customStruct)(nil),
	)
	t.Logf(b.String())
}

```

- The `case time.Time` was never hit (handled earlier in formatLogfmtValue)
- The `nil-concrete-stringer` is the only one that triggers a panic, and where we need the `recover`. 

Also there's this wonky case of `nil` vs `<nil>` (not  your fault, been like this since a long way back): 
```
    logger_test.go:50: INFO [11-09|08:14:53.459] test                                     nil-stringer-interface=nil nil-error-interface=nil nil-custom-err= nil-concrete-stringer=nil nil-custom-struct=<nil>
```

holiman replied:
It used to be that `formatShared` was called either from `formatJSONValue` or `formatLogfmtValue`. Now the former is removed, we can inline it into `formatLogfmtValue`. That makes a lot ot sense: both `formatLogfmtValue` and `formatShared` are just type-switches on `v`, so it's a lot better to merge them.

holiman replied:
As far as I can make out, the `formatShared` can essentially be reduced into 

```golang
func formatShared(value interface{}) (result interface{}) {
	switch v := value.(type) {
	case error:
		return v.Error()

	case fmt.Stringer:
		if v := reflect.ValueOf(v); v.Kind() == reflect.Ptr && v.IsNil() {
			return "caughtnil"
		}
		return v.String()
	default:
		return v
	}
}
```
(and then inlined above)

holiman commented on cmd/geth/logtestcmd_active.go:
```diff
@@ -130,5 +130,9 @@ func logTest(ctx *cli.Context) error {
 		log.Info("Inserted known block", "number", 99, "hash", common.HexToHash("0x12322"), "txs", 10, "gas", 1, "other", "third")
 		log.Warn("Inserted known block", "number", 1_012, "hash", common.HexToHash("0x1234"), "txs", 200, "gas", 99, "other", "fourth")
 	}
+	{
+		log.Info("nil-big.Int", "<nil>", (*big.Int)(nil))
+		log.Info("nil-uint256.Int", "<nil>", (*uint256.Int)(nil))
+	}
```
it's better if we commit this to master first, see https://github.com/ethereum/go-ethereum/pull/28501

holiman commented on log/handler_glog.go:
```diff
@@ -130,107 +126,77 @@ func (h *GlogHandler) Vmodule(ruleset string) error {
+}
+
+func (h *GlogHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
+	res := GlogHandler{
+		h.origin.WithAttrs(attrs),
+		atomic.Int32{},
+		atomic.Bool{},
+		h.patterns,
+		h.siteCache,
+		h.location,
+		sync.RWMutex{},
```
First of all, I think it would be nicer to rewrite it thus: 
```
	res := GlogHandler{
		origin:    h.origin.WithAttrs(name),
		patterns:  h.patterns,
		siteCache: h.siteCache,
		location:  h.location,
	}
```

Secondly, I don't think this is safe. This returns a new `GlogHandler`, which uses the same instances of `h.siteCache` and `h.patterns`, but guarded by a new and different lock. 

So I think you need to obtain the lock, then copy patthers/sitecache and hand over to the new instance.

holiman commented on log/handler_glog.go:
```diff
@@ -130,107 +126,77 @@ func (h *GlogHandler) Vmodule(ruleset string) error {
+}
+
+func (h *GlogHandler) WithGroup(name string) slog.Handler {
+	panic("not implemented")
```
Please implement

jwasinger replied:
I'm working on adding support for groups but it's going to be somewhat of a nontrivial change.

jwasinger replied:
reason it seems nontrivial is that an `Attribute` can be a group, which itself can contain subgroups.

holiman replied:
Ok, we can save that for a later PR

holiman commented on log/handler.go:
```diff
@@ -1,375 +1,290 @@
+// TerminalHandlerWithLevel returns the same handler as TerminalHandler but only outputs
+// records which are less than or equal to the specified verbosity level.
+func TerminalHandlerWithLevel(wr io.Writer, lvl slog.Level, useColor bool) slog.Handler {
+	return &terminalHandler{
+		sync.Mutex{},
+		wr,
+		lvl,
+		useColor,
+		[]slog.Attr{},
+	}
```
How about this?
```suggestion
		return &terminalHandler{
		wr:       wr,
		lvl:      lvl,
		useColor: useColor,
	}
```
(but make sure the `nil` instead of empty attr list works)

holiman commented on log/handler.go:
```diff
@@ -1,375 +1,290 @@
+func (h *terminalHandler) Handle(_ context.Context, r slog.Record) error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	r.AddAttrs(h.attrs...)
+	h.wr.Write(TerminalFormat(r, h.useColor))
```
Might be better to pass the `h.attrs` to `TerminalFormat`, instead of temporarily adding them to the record. The `TerminalFormat` internally calls `logfmt`, which _does_ have a parameter `commonAttrs []slog.Attr` where these extra attrs would fit perfectly (?)

jwasinger replied:
Yeah good idea.

jwasinger replied:
weird.  that extra parameter is unused anywhere.  the code that used it must have been deleted at some point...

holiman replied:
Yep, when you dropped `LogfmtFormat()`:

> If you also drop the method LogfmtFormat(), then TerminalFormat is the only caller of logfmt, so that could also be made into a function on terminalHandler, and suddenly both the fieldPadding and the handler's attrs are accessible for both these methods.

holiman commented:
One thing to consider: should we leave everything under `log` intact (with deprecation-notice), put all the new things under e.g. `slog`? Then our own code could use `slog` exclusively, but downstream projects would be unaffected. 

Or maybe just let all the old stuff live under `legacylog`? Then downstreams would detect the problem, and have a choice to either do a very simple short-term fix (change `go-ethereum/log` to `go-ethereum/legacylog`) or do the more thorough proper change. 

cc @fjl

holiman commented:
If we don't have to care about backwards compatiblity API-wise (which we've already thrown overboard), then IMO we could make `TerminalFormat` non-public, and maybe be a member method on `terminalHandler`. Doing that would allow us to use a `fieldPadding` map which is owned by the `terminalHandler`.

holiman commented on log/handler.go:
```diff
@@ -1,375 +1,228 @@
+func builtinReplace(_ []string, attr slog.Attr) slog.Attr {
+	switch attr.Key {
+	case slog.TimeKey:
+		if t, ok := attr.Value.Any().(time.Time); ok {
+			attr = slog.Any("t", t.Format(timeFormat))
+			return attr
+		}
```
```suggestion
		if attr.Value.Kind() == slog.KindTime {
			return slog.Time("t", attr.Value.Time())
		}
```

holiman commented on log/handler.go:
```diff
@@ -170,104 +170,42 @@ func (l *leveler) Level() slog.Level {
 
 func JsonHandler(wr io.Writer) slog.Handler {
-	return &builtinHandler{slog.NewJSONHandler(wr, &slog.HandlerOptions{
+	return slog.NewJSONHandler(wr, &slog.HandlerOptions{
```
Hm,this `JsonHandler` method is not used, our code uses
```
		handler = slog.NewJSONHandler(output, nil)
```
Which means that geth doesn't actually use the ReplaceAttr functionality for json at all,

jwasinger replied:
oof

jwasinger commented:
@holiman Your [test case](https://github.com/ethereum/go-ethereum/blob/master/cmd/geth/logtestcmd_active.go#L141) found a bug/crash in slog's logfmt handler:

```
t=2023-11-09T19:06:18+0800 lvl=info msg=(fmt.Stringer)(nil) res=<nil>
panic: value method time.Time.MarshalText called using nil *Time pointer

goroutine 1 [running]:
time.(*Time).MarshalText(0x14000a067d8?)
        <autogenerated>:1 +0x48
golang.org/x/exp/slog.appendTextValue(0x14000a06f48, {{}, 0x10602b160?, {0x10602b160?, 0x0?}})
        /Users/jwasinger/go/pkg/mod/golang.org/x/exp@v0.0.0-20230905200255-921286631fa9/slog/text_handler.go:104 +0x280
golang.org/x/exp/slog.(*handleState).appendValue(0x14000a06f48, {{}, 0x1056e348c?, {0x10602b160?, 0x0?}})
        /Users/jwasinger/go/pkg/mod/golang.org/x/exp@v0.0.0-20230905200255-921286631fa9/slog/handler.go:511 +0x3c
golang.org/x/exp/slog.(*handleState).appendAttr(0x14000a06f48, {{0x1056e348c, 0x3}, {{}, 0x0, {0x10602b160, 0x0}}})
        /Users/jwasinger/go/pkg/mod/golang.org/x/exp@v0.0.0-20230905200255-921286631fa9/slog/handler.go:467 +0x6fc
golang.org/x/exp/slog.(*handleState).appendNonBuiltIns.func1({{0x1056e348c, 0x3}, {{}, 0x0, {0x10602b160, 0x0}}})
        /Users/jwasinger/go/pkg/mod/golang.org/x/exp@v0.0.0-20230905200255-921286631fa9/slog/handler.go:320 +0x4c
golang.org/x/exp/slog.Record.Attrs({{0xc14b4d2ab5369ea0, 0x1c3b1f4, 0x106d219c0}, {0x105701d92, 0x15}, 0x0, 0x10560e708, {{{0x1056e348c, 0x3}, {{...}, ...}}, ...}, ...}, ...)
        /Users/jwasinger/go/pkg/mod/golang.org/x/exp@v0.0.0-20230905200255-921286631fa9/slog/record.go:84 +0x98
golang.org/x/exp/slog.(*handleState).appendNonBuiltIns(_, {{0xc14b4d2ab5369ea0, 0x1c3b1f4, 0x106d219c0}, {0x105701d92, 0x15}, 0x0, 0x10560e708, {{{0x1056e348c, 0x3}, ...}, ...}, ...})
        /Users/jwasinger/go/pkg/mod/golang.org/x/exp@v0.0.0-20230905200255-921286631fa9/slog/handler.go:319 +0x3d8
golang.org/x/exp/slog.(*commonHandler).handle(_, {{0xc14b4d2ab5369ea0, 0x1c3b1f4, 0x106d219c0}, {0x105701d92, 0x15}, 0x0, 0x10560e708, {{{0x1056e348c, 0x3}, ...}, ...}, ...})
        /Users/jwasinger/go/pkg/mod/golang.org/x/exp@v0.0.0-20230905200255-921286631fa9/slog/handler.go:297 +0x450
golang.org/x/exp/slog.(*TextHandler).Handle(_, {_, _}, {{0xc14b4d2ab5369ea0, 0x1c3b1f4, 0x106d219c0}, {0x105701d92, 0x15}, 0x0, 0x10560e708, ...})
        /Users/jwasinger/go/pkg/mod/golang.org/x/exp@v0.0.0-20230905200255-921286631fa9/slog/text_handler.go:93 +0x44
github.com/ethereum/go-ethereum/log.(*GlogHandler).Handle(_, {_, _}, {{0xc14b4d2ab5369ea0, 0x1c3b1f4, 0x106d219c0}, {0x105701d92, 0x15}, 0x0, 0x10560e708, ...})
        /Users/jwasinger/projects/go-ethereum-slog/log/handler_glog.go:172 +0x264
```

holiman commented:
Yes, I noticed it too :)

holiman commented:
If you remove the replaceAttr, the crash happens earlier, on the nil `uint256` . Same place though : 
```
	case KindAny:
		if tm, ok := v.any.(encoding.TextMarshaler); ok {
			data, err := tm.MarshalText()
```

holiman commented:
https://github.com/golang/go/issues/64034

holiman commented:
The crash is fixed in `log/slog`, on golang@dev (not 1.21), but it's not fixed in `golang.org/x/exp/slog` yet.

holiman commented:
I'd call this a blocker. I don't want geth to crash just because we log some object which is `nil` and happens to implement Marshaller or Stringer (in the specific case I reported, the stringer was fine, but in most cases evaluation of `x.String()` will panic too). 
Once we get the fix merged into `x/exp/slog`, we can update the dependency and move on.

jwasinger commented on cmd/geth/logtestcmd_active.go:
```diff
@@ -42,7 +42,6 @@ This command is only meant for testing.
 // logTest is an entry point which spits out some logs. This is used by testing
 // to verify expected outputs
 func logTest(ctx *cli.Context) error {
-	log.ResetGlobalState()
```
Note to self:  create and set a new default logger here to clear the field padding map.

jwasinger commented on p2p/discover/v5_udp_test.go:
```diff
@@ -82,13 +79,7 @@ func startLocalhostV5(t *testing.T, cfg Config) *UDPv5 {
-		t.Logf("%s %s", lprefix, log.TerminalFormat(r, false))
-		return nil
-	}))
+	cfg.Log = testlog.Logger(t, log.LevelTrace).With("node-id", lprefix)
```
This change is necessary because `TerminalFormat` is now a private member method of `terminalHandler`.

It has the downside of making it harder to read the output of the tests because the `node-id` is not the leftmost item on the output lines.

jwasinger commented:
whoops, I applied https://github.com/ethereum/go-ethereum/pull/20528 here and will revert that.

holiman commented on cmd/geth/logtestcmd_active.go:
```diff
@@ -42,6 +44,11 @@ This command is only meant for testing.
 func logTest(ctx *cli.Context) error {
+	// if log is set to output terminal format, reset root logger to clear field padding map
+	if !ctx.IsSet(debug.LogFormatFlag.Name) || ctx.String(debug.LogFormatFlag.Name) == "terminal" {
+		log.SetDefault(log.NewLogger(log.TerminalHandler(os.Stdout, false)))
```
I changed it, so that internal/debug has access to the terminal format, but thne I had to make that public, and since it clashed with the constructur _and_ the constructor was named `TerminalHandler` instead of the more go-idiomatic `NewTerminalHandler`, I renamed the ctor and touched ~20+ files. Sigh. Anyway, feel free to undo my change if you don't like it, or think of something better 

Now I get no more errors on the `terminal`, but it appears the that the `logfmt` suddenly has timezone info in the timestamp, which apparently it didn't have before.

holiman replied:
Ah no it wasn't the tz that was new: 
```
old t=2023-10-20T12:56:08+0200
new t=2023-11-13T20:51:45.938+01:00
```
So... we should ignore millisecond level. I dunno, is the old (second-resolution) better than millisecond-resolution? For a machine-readable format???  Meh....

holiman replied:
Hm, actually two diffs: 
- milliseconds `.938`
- different format for tz, `+0200` vs `+01:00`.

jwasinger replied:
The built-in time field stopped being formatted according to our current time format at some point.

jwasinger replied:
crap we lost comma formatting for `*.Int`s somewhere along the way.

holiman replied:
No that's fine: https://github.com/ethereum/go-ethereum/pull/28187#issuecomment-1801291049

holiman commented:
I took a peek at the `json` outout too. 
New slog logger: `,"<nil>":"<nil>"`
Old logger : "\u003cnil\u003e":"\u003cnil\u003e"

So the old logger saw fit to use unicode escape for `<` and `>`. Is it correct? According to this SO poster, https://stackoverflow.com/a/28885991 , 

> As explained in the section 9 of the official ECMA specification (https://www.ecma-international.org/wp-content/uploads/ECMA-404_2nd_edition_december_2017.pdf) ...

It seems correct to unicode-escape them. Which would mean that the old logger was more conformant. And that is indeed how golang `json` marshaller treats it: 
```
func TestFoo(t *testing.T) {
	a, _ := json.Marshal(map[string]string{"<nil>": "<nil>"})
	t.Log(string(a))
}
```
```
    format_test.go:27: {"\u003cnil\u003e":"\u003cnil\u003e"}
```

holiman commented:
Seems the `slog` people intentionally disabled it
```
// Modified from encoding/json/encode.go:encodeState.string,
// with escapeHTML set to false.
func appendEscapedJSONString(buf []byte, s string) []byte {
```
Oh well, I guess we can live with that (?)

holiman commented on log/handler.go:
```diff
@@ -177,7 +177,7 @@ func builtinReplace(_ []string, attr slog.Attr) slog.Attr {
 	case slog.TimeKey:
 		if attr.Value.Kind() == slog.KindTime {
-			return slog.Time("t", attr.Value.Time())
+			return slog.Any("t", attr.Value.Time().Format(timeFormat))
```
Why use `Any` when you know it's a `String` ?

holiman commented on cmd/geth/testdata/logging/logtest-terminal.txt:
```diff
@@ -1,50 +1,50 @@
-INFO [XX-XX|XX:XX:XX.XXX] raw nil                                  res=nil
-INFO [XX-XX|XX:XX:XX.XXX] (*uint64)(nil)                           res=<nil>
-INFO [XX-XX|XX:XX:XX.XXX] Using keys 't', 'lvl', 'time', 'level' and 'msg' t=t time=time lvl=lvl level=level msg=msg
+INFO [11-14|22:14:53.479] big.Int                                  111,222,333,444,555,678,999=111,222,333,444,555,678,999
```
Other than that, I think all the changes in these files are ok.

holiman commented on internal/debug/flags.go:
```diff
@@ -167,9 +167,14 @@ var Flags = []cli.Flag{
 var (
 	glogger                *log.GlogHandler
-	logOutputF             *os.File
+	logOutputF             writerCloser
```
I know I could have said something earlier, but.. `logOutputF` is not a great name. Can we make it a bit more descripitive? Like `logOutputFile` `logFile` or `logWriter`

Also, why not just use `io.WriteCloser`: https://pkg.go.dev/io#WriteCloser ?

holiman commented on log/format.go:
```diff
@@ -177,14 +151,18 @@ func FormatLogfmtValue(value interface{}, term bool) string {
+	case error:
+		return v.Error()
+
+	case fmt.Stringer:
+		return v.String()
```
So, I suspect that `v` being nil here might `panic`, and we might have to add back that `recover` clause, _or_ check for nil (the tests should tell us). 

However, it's usually not nil, so it might be preferrable to have an expensive recover rather than regularly check for nilness. At least it seems that the implementors of slog have chosen that route.

holiman replied:
Yup

```
$ go run -tags=integrationtests ./cmd/geth --log.format terminal logtest  
INFO [11-14|20:16:15.800] big.Int                                  111,222,333,444,555,678,999=111,222,333,444,555,678,999
...
INFO [11-14|20:16:15.802] (fmt.Stringer)(nil)                      res=nil
panic: value method time.Time.String called using nil *Time pointer

goroutine 1 [running]:
time.(*Time).String(0x0?)
```

holiman replied:
Ah, you also moved the `stringer` earlier, so now the `common.Hash` triggers the `String()` instead of `TerminalString()`. I'll push a fix

holiman commented:
Rebased to fix go.mod-conflicts

holiman commented on log/format.go:
```diff
@@ -375,8 +167,11 @@ func formatLogfmtValue(value interface{}, term bool) string {
 	}
-	value = formatShared(value)
 	switch v := value.(type) {
+	case error:
+		return v.Error()
+	case fmt.Stringer:
+		return v.String()
```
Wait. Both of these needs `escapedString`, right? We should add testcase (on `master`) which logs:

- an error with spaces `errors.New("error: an 'error' occurred")` and 
- a `fmt.Stringer` which also spits out text spaces (needs quoting)

holiman commented on log/handler_glog.go:
```diff
@@ -130,107 +126,84 @@ func (h *GlogHandler) Vmodule(ruleset string) error {
+	if slog.Level(h.level.Load()) <= r.Level {
+		return h.origin.Handle(context.Background(), r)
 	}
+
```
We should keep this, no? Seems like a valid optimization to me
 
```suggestion
	if !h.override.Load() {
		return nil
	}
```

jwasinger replied:
This is checked in `Enabled`.

holiman reviewed (approved):
Looks good to me! 
Before we merge, just a couple of things:

1.  We should do a sync (sanity check on the log output)

2. We should consider some options to make life simper for projects using `go-ethereum/log` as a library. Some options that I can think of are:
   1. Put our `slog` stuff into `go-ethereum/slog`, and never touch `log` (add deprecation notices though). Easy for downstreams (they probably won't even notice). 
   3. Put our `slogs` stuff into either `go-ethereum/slog` or `go-ethereum/log`, but move the old stuff into `go-ethereum/legacylog`. Dependent projects will break, but we can just tell them to repoint the dependency from `log` to `legacylog`. There's even some go alias magic we can make use of. 
   4. Put our `slogs` stuff into either `go-ethereum/slog` or `go-ethereum/log`, but put our old stuff into a (dead) new repo: e.g. `ethereum/geth-legacy/log`. Dependent projects will break, but we can just tell them to repoint the dependency to the new repository. (Possibly use alias magic?)

I don't think it's good to do what we do right now, which just overwrites the `log` package in a non-backwards compatible way with no easy way to un-break. I mean, we don't want people to not be able to upgrade their geth-dependency.

holiman commented:
Ok, there are a couple of more polishes that we should consider for `json`. 

### Timestamp

- Old format uses timestamp `2023-11-09T08:33:19.465673664+01:00`. 
- New format uses timestamp `2023-11-16T20:58:02+0100`

This becomes fixed if the json `builtinReplace` simply does
```golang
	switch attr.Key {
	case slog.TimeKey:
		return slog.Attr{attr.Key, attr.Value}
```
instead of formatting. It means we'd have to split builtinReplace and not reuse between json/logfmt. 

### Nil handling

The new format actually convert some types of `nil` into json `null`. I think this is good, but it _is_ a change. 
```

{"time":"2023-11-16T21:02:09.794626967+01:00","lvl":"info","msg":"(fmt.Stringer)(nil)","res":null}
{"time":"2023-11-16T21:02:09.794657219+01:00","lvl":"info","msg":"nil-concrete-stringer","res":null}
{"time":"2023-11-16T21:02:09.794700283+01:00","lvl":"info","msg":"error(nil) ","res":null}
{"time":"2023-11-16T21:02:09.794751672+01:00","lvl":"info","msg":"nil-custom-struct","res":null}
{"time":"2023-11-16T21:02:09.794840389+01:00","lvl":"info","msg":"raw nil","res":null}
{"time":"2023-11-16T21:02:09.794867452+01:00","lvl":"info","msg":"(*uint64)(nil)","res":null}
```
Some `nil`s are not `null`:
```
{"time":"2023-11-16T21:02:09.7935443+01:00","lvl":"info","msg":"(*big.Int)(nil)","<nil>":"<nil>"}
{"time":"2023-11-16T21:02:09.794584148+01:00","lvl":"info","msg":"(*uint256.Int)(nil)","<nil>":"<nil>"}
{"time":"2023-11-16T21:02:09.794725075+01:00","lvl":"info","msg":"nil-concrete-error","res":""}
```

The old format had IMO pretty ugly nil-cases:
```
{"\u003cnil\u003e":"\u003cnil\u003e","lvl":"info","msg":"(*big.Int)(nil)","t":"2023-11-09T08:33:19.465450283+01:00"}
{"\u003cnil\u003e":"nil","lvl":"info","msg":"(*uint256.Int)(nil)","t":"2023-11-09T08:33:19.465472953+01:00"}
{"lvl":"info","msg":"(fmt.Stringer)(nil)","res":"\u003cnil\u003e","t":"2023-11-09T08:33:19.465538633+01:00"}
{"lvl":"info","msg":"nil-concrete-stringer","res":"nil","t":"2023-11-09T08:33:19.465552355+01:00"}
{"lvl":"info","msg":"error(nil) ","res":"\u003cnil\u003e","t":"2023-11-09T08:33:19.465601029+01:00"}
{"lvl":"info","msg":"nil-concrete-error","res":"","t":"2023-11-09T08:33:19.46561622+01:00"}
{"lvl":"info","msg":"nil-custom-struct","res":"\u003cnil\u003e","t":"2023-11-09T08:33:19.465638888+01:00"}
{"lvl":"info","msg":"raw nil","res":"\u003cnil\u003e","t":"2023-11-09T08:33:19.465673664+01:00"}
{"lvl":"info","msg":"(*uint64)(nil)","res":"\u003cnil\u003e","t":"2023-11-09T08:33:19.465700264+01:00"}
```
I like that the new json output is actually pretty readable: it outputs time, level on the left side, making the columns align. 

----

I'm fine with the `json` changes re nil, but maybe we should retain the previous (higher-resolution) timestamp.

jwasinger commented:
> We should consider some options to make life simper for projects using go-ethereum/log as a library. Some options that I can think of are:

1 seems like the easiest option here for us.

jwasinger commented:
The `logfmt`/`json` stringer mismatch is a weird one because `fmt.Stringer` is not referenced anywhere in the `slog` codebase.

jwasinger commented:
> The `logfmt`/`json` custom stringer type mismatch is a weird one because `fmt.Stringer` is not referenced anywhere in the `slog` codebase.

it's because `json.Encoder` doesn't handle `fmt.Stringer` properly: https://go.dev/play/p/hMn_JAb8eRx

jwasinger commented:
Would it be worth adding json tests which parse each line from the output/testdata, and make sure the same key/value pairs are in both (ignoring/`x`ing time of course)?

holiman commented:
> Would it be worth adding json tests which parse each line from the output/testdata,

Yeah, I do. Just soup them up as a `[string]any` map, I guess that would do the trick.

holiman commented:
Marking for triage to discusss https://github.com/ethereum/go-ethereum/pull/28187#pullrequestreview-1735335758

holiman commented:
This is now running on `bench05`, master at `bench06` for comparison

jwasinger commented:
Do we want to have `LogAttrs`, `TraceAttrs`, `WarnAttrs`, etc...?  I haven't done investigation to the performance gain by switching over to moving all the log call sites to typed attributes.  But it should be decided before we finalize a public interface for the logger.

Even if it doesn't result in substantial gain for us, maybe it will be useful downstream? adding it to this PR shouldn't make it much larger.  It just adds a larger API surface to the logger.

holiman commented:
> Do we want to have `LogAttrs`, `TraceAttrs`, `WarnAttrs`, etc...? I haven't done investigation to the performance gain by switching over to moving all the log call sites to typed attributes. But it should be decided before we finalize a public interface for the logger.
> 
> Even if it doesn't result in substantial gain for us, maybe it will be useful downstream? adding it to this PR shouldn't make it much larger. It just adds a larger API surface to the logger.


Probably, but I don't see any need to add it from the get-go. We could add it in a follow-up. Do we want to continue using `log.TraceAttrs`, or should callsites that wish to adopt this pattern instead use `slog.TraceAttrs`? Because if we're going to rewrite the callsites, might aswell rewrite it fully and avoid `log`. Unless there's some compelling reason to keep it as it is. 

Anyway, we don't have to decide it here and now (actually, I'd prefer it if we postpone it for later)

fjl commented:
I think we should not add too many wrappers. Most code should probably just switch over to using `slog` eventually.

namiloh commented on log/root.go:
```diff
@@ -2,22 +2,26 @@ package log
-	root.Store(l)
+	rootMu.Lock()
+	defer rootMu.Unlock()
+	root := l
```
This redefines root. Did you mean '=' not ':=' ?

jwasinger replied:
Yeah good catch.

jwasinger commented on log/root.go:
```diff
@@ -26,8 +30,10 @@ func Root() Logger {
 func rootLogger() *logger {
-	res, _ := root.Load().(*logger)
-	return res
+	rootMu.Lock()
```
I use a mutex because I wanted to guarantee that concurrent calls to `log.SetDefault` wouldn't result in the slog root logger and `root` having inconsistent values.

namiloh replied:
Imo should not be much of a problem. Should not be massively invoked

jwasinger commented:
Actually `98166316b918ef1ad5e431aa10f08c6ee2707a19` probably needs to be reverted.  I have to think about how to approach this.  Maybe enforce that the logger passed to `log.SetDefault` is a `log.logger`.

namiloh commented:
> Actually `98166316b918ef1ad5e431aa10f08c6ee2707a19` needs to be reverted. I have to think about how to approach this. Maybe enforce that the logger passed to `log.SetDefault` is a `log.logger`.

That is already enforced by the method signature. How do you mean?

jwasinger commented:
the method requires a `Logger`, not a `logger`.  exposing a `Write` method on the `Logger` object would allow everything in `root.go` to be written using `Logger` instead of `logger`.

jwasinger commented:
> Anyway, we don't have to decide it here and now (actually, I'd prefer it if we postpone it for later)

If we define a public interface, we can't add methods without breaking backwards compatibility.  so we should probably make a decision.

namiloh commented on log/root.go:
```diff
@@ -2,42 +2,30 @@ package log
-	rootMu.Lock()
-	defer rootMu.Unlock()
-	root := l
+	root = l
```
I would go with atomic, like you did before, and like slog does https://cs.opensource.google/go/x/exp/+/master:slog/logger.go;l=30

holiman commented on log/root.go:
```diff
@@ -2,31 +2,32 @@ package log
 // Root returns the root logger
 func Root() Logger {
-	return root
+	res, _ := root.Load().(Logger)
+	return res
```
```suggestion
	return root.Load().(Logger)
```

jwasinger commented:
I'm just trying to finish this up by adding equivalence testing for the json output.

So we are okay with omitting log `*Attrs` methods for now? and not providing an [interface method](https://github.com/ethereum/go-ethereum/blob/77b6db7b4587eb278279cda424f50b3d11a3ad59/log/logger.go#L108) to instantiate a `log.Logger` with group(s)?

Potentially adding these in the future could break backwards compatibility.

holiman commented:
I have changed by mind about `Inner()`. It seemed to me that we were still leaking internals, when we had a `Handler()`-method to retrieve the `inner.Hander`. But looking at it now, I don't see why we need to expose any of it. 
The SetDefault can look like this: 
```golang
func SetDefault(l Logger) {
	root.Store(l)
	if lg, ok := l.(*logger); ok {
		slog.SetDefault(lg.inner)
	}
}
```
And we can just drop `Inner` from the interface. It compiles, not sure what I'm missing...

holiman commented:
> So we are okay with omitting log *Attrs methods for now?

Yes.

> and not providing an [interface method](https://github.com/ethereum/go-ethereum/blob/77b6db7b4587eb278279cda424f50b3d11a3ad59/log/logger.go#L108) to instantiate a log.Logger with group(s)?

I think so, yes... (?)

jwasinger commented:
... and we would be inhibited from adding these in the future because of breaking compatibility with the interface we are about to introduce with this PR right?

jwasinger commented:
> I have changed by mind about `Inner()`. It seemed to me that we were still leaking internals, when we had a `Handler()`-method to retrieve the `inner.Hander`. But looking at it now, I don't see why we need to expose any of it. The SetDefault can look like this:
> 
> ```go
> func SetDefault(l Logger) {
> 	root.Store(l)
> 	if lg, ok := l.(*logger); ok {
> 		slog.SetDefault(lg.inner)
> 	}
> }
> ```
> 
> And we can just drop `Inner` from the interface. It compiles, not sure what I'm missing...

The reason for inner, would be to allow setting the root logger as a `Logger` that isn't necessarily a `logger`.  There's no compelling use-case to include this for us.  And we don't really care to support overriding the root logger with a custom `Logger` by users.  So sure, we can remove this and do what you are suggesting.

holiman commented:
> ... and we would be inhibited from adding these in the future because of breaking compatibility with the interface we are about to introduce with this PR right?

More or less, yes. But afaict, if someone _wants_ to use `WithAttrs`, then they can just use `slog` directly, not go via `go-ethereum/log`, right?

jwasinger commented:
> I have changed by mind about `Inner()`. It seemed to me that we were still leaking internals, when we had a `Handler()`-method to retrieve the `inner.Hander`. But looking at it now, I don't see why we need to expose any of it. The SetDefault can look like this:
> 
> ```go
> func SetDefault(l Logger) {
> 	root.Store(l)
> 	if lg, ok := l.(*logger); ok {
> 		slog.SetDefault(lg.inner)
> 	}
> }
> ```
> 
> And we can just drop `Inner` from the interface. It compiles, not sure what I'm missing...

Deleted my incorrect comment.  Yeah we don't need `Inner`.  If a user sets the default logger with something that isn't a `logger`, this will silently cause any log invocations that are made using `slog` to silently not appear.  But I don't think this is a huge problem.

holiman commented:
> 
> Deleted my incorrect comment. Yeah we don't need `Inner`. If a user sets the default logger with something that isn't a `logger`, this will silently cause any log invocations that are made using `slog` to silently not appear. But I don't think this is a huge problem.

I didn't see anything incorrect, but yeah, the only fallout will be that a custom logger won't 'stick' on both `log` and `slog` invocations equally. Such a caller can fix it by invoking `SetDefault` both on `log` and `slog` though.

jwasinger commented:
> > ... and we would be inhibited from adding these in the future because of breaking compatibility with the interface we are about to introduce with this PR right?
> 
> More or less, yes. But afaict, if someone _wants_ to use `WithAttrs`, then they can just use `slog` directly, not go via `go-ethereum/log`, right?

Oh right. yeah.

jwasinger commented:
The json equivalence tester is actually a bit annoying.  The plan was to go line-by-line comparing test data with received output, and ignoring the time.  And this would be simple enough because I can easily transform each line string into a `map[string][]byte`.

However, in cases where user-provided keys clash with the keys used by record object built-in attributes, slog produces invalid json.  So writing the tester is a bit hackier (have to "decode" each line by extracting fields/values directly from the string).  still straightforward enough I think.

holiman commented:
> The json equivalence tester is actually a bit annoying. The plan was to go line-by-line comparing test data with received output, and ignoring the time. And this would be simple enough because I can easily transform each line string into a `map[string][]byte`.

I think `map[string]any` might be better.
 
> However, in cases where user-provided keys clash with the keys used by record object built-in attributes, slog produces invalid json with duplicate keys.

It's not so invalid that the golang  json parser refuses to parse it, so let's just do that. That's what a "consumer" of this json stream would do, so that's IMO the best way to test it

holiman commented:
```golang
func TestJsonLogging(t *testing.T) {
	t.Parallel()
	haveB, err := runSelf("--log.format", "json", "logtest")
	if err != nil {
		t.Fatal(err)
	}
	readFile, err := os.Open("testdata/logging/logtest-json.txt")
	if err != nil {
		t.Fatal(err)
	}
	wantLines := split(readFile)
	haveLines := split(bytes.NewBuffer(haveB))
	for i, want := range wantLines {
		if i > len(haveLines)-1 {
			t.Fatalf("format %v, line %d missing, want:%v", "json", i, want)
		}
		have := haveLines[i]
		for strings.Contains(have, "Unknown config environment variable") {
			// This can happen on CI runs. Drop it.
			haveLines = append(haveLines[:i], haveLines[i+1:]...)
			have = haveLines[i]
		}
		var h map[string]any
		var w map[string]any
		if err := json.Unmarshal([]byte(have), &h); err != nil {
			t.Fatal(err)
		}
		if err := json.Unmarshal([]byte(want), &w); err != nil {
			t.Fatal(err)
		}
		h["t"] = "xxx"
		w["t"] = "xxx"
		outh, _ := json.Marshal(h)
		outw, _ := json.Marshal(w)
		if !bytes.Equal(outh, outw) {
			// show an intelligent diff
			t.Logf(nicediff(outh, outw))
			t.Errorf("file content wrong")
		}
	}
}
```
Works for me ... ?

jwasinger commented:
From RFC 8529:
> The names within an object SHOULD be unique.

So I guess it's not strictly required and may deviate between software implementations.  Okay, then checking that `testdata == expected` according to the Go json marshaller is probably the proper thing to do here.  Pls push that commit to the PR.

holiman commented:
```
>>> build/cache/golangci-lint-1.51.1-linux-amd64/golangci-lint run --config .golangci.yml ./...
log/handler.go:188:12: composites: golang.org/x/exp/slog.Attr struct literal uses unkeyed fields (govet)
				return slog.Attr{"t", attr.Value}
				       ^
util.go:48: exit status 1
```

holiman reviewed (approved):
LGTM

islishude commented:
the pr prevents all logs

```
package main

import (
	"context"
	"fmt"
	"log"

	"github.com/ethereum/go-ethereum/ethclient"
)

func main() {
	rpcclient, err := ethclient.Dial("https://eth.llamarpc.com")
	if err != nil {
		panic(err)
	}

	fmt.Println("...log.Println has no outpus")
	log.Println(rpcclient.ChainID(context.TODO()))
	fmt.Println("...log.Println has no outpus")
}
```

Can we remove this line?

https://github.com/ethereum/go-ethereum/blob/09e0208029ff96a9cda0c69dbaebfd3f31a39771/log/root.go#L14-L15

![image](https://github.com/ethereum/go-ethereum/assets/24730006/17a8fd8f-1f6b-483a-b5d6-563bd90b96f4)

fjl commented:
https://github.com/ethereum/go-ethereum/issues/28738 is about this problem
